### PR TITLE
Sorting tables

### PR DESCRIPTION
**Pull request by robcar**
Adding possibility to sort contents of a table

### CODE REVIEW

File: src/components/elements/table/header.js [hunk 1]
```diff
@@ -1,21 +1,53 @@
...
-export default ({ api, property }) => {
-  if (!property.columns) {
-    return <thead/>
+export default class Header extends Component {
+  constructor(props) {
+    super()
+
+    this.state = {
+      sortedColumn: null,
+      direction: null,
```
Comment (nranas) [inline src/components/elements/table/header.js]: Maybe this property should be named `descending` instead considering how it is used.

File: src/components/elements/table/header.js [hunk 2]
```diff
@@ -1,21 +1,53 @@
...
+          })
+          }
+        </tr>
+      </thead>
+    )
+  }
+
+  onSort(e, columnIndex) {
+    e.preventDefault()
+    this.setState({
```
Comment (nranas) [inline src/components/elements/table/header.js]: I'm not sure if you think it would be a good idea or not but maybe we could keep all sorting state at the level above this one (`index.js`) and send the sorted column index and direction as properties to this component. Then we could keep this component stateless.

File: src/components/elements/table/header.js [hunk 3]
```diff
@@ -1,21 +1,53 @@
...
+  }
+
+  render() {
+    if (!this.props.columns) {
+      return <thead />
+    }
+    return (
+      <thead>
+        <tr className='ct-table-header-row'>
+          {this.props.columns.map((column, columnIndex) => {
```
Comment (hfjallemark) [inline src/components/elements/table/header.js]: No need for `{` and `return`:

```
{this.props.columns.map((column, columnIndex) => (
  <th>
    ...
  </th>
)}

File: src/components/elements/table/header.js [hunk 4]
```diff
@@ -1,21 +1,53 @@
+import { Component } from 'react'
 import cx from 'classnames'
 
-export default ({ api, property }) => {
-  if (!property.columns) {
-    return <thead/>
+export default class Header extends Component {
+  constructor(props) {
```
Comment (hfjallemark) [inline src/components/elements/table/header.js]: I believe we can use [class properties](https://babeljs.io/docs/plugins/transform-class-properties) instead of setting `state` in the constructor (that way we don't need to add a constructor and remember to call super, etc), like so:

```
export default class Header extends Component {
  state = {
    sortedColumn: null,
    direction: null,
  }
}
```

File: src/components/elements/table/header.js (hunk 2 above)
Comment (hfjallemark) [reply src/components/elements/table/header.js]: Agree, that probably makes more sense since the level above needs to know it anyway.

File: src/components/elements/table/index.js [hunk 5]
```diff
@@ -1,14 +1,67 @@
...
+    })
+    return false
+  }
+
+  getRowValue(row, columnIndex) {
+    if (row[columnIndex].type == 'html') {
+      const doc = document.implementation.createHTMLDocument('')
+      doc.documentElement.innerHTML = row[columnIndex].content
+
+      return doc.getElementsByTagName('body')[0].innerText
```
Comment (hfjallemark) [inline src/components/elements/table/index.js]: This is pretty smart! If performance is an issue, one alternative could be to just strip out the tags with a regex (`<[^>]*>`, the way @npehrsson would have liked it!).

File: src/components/elements/table/header.js (hunk 2 above)
Comment (robcar) [reply src/components/elements/table/header.js]: I was unsure what you guys would prefer I thought you might want to keep state local to each component, that is why I did it :)
But it makes sense to keep state in just one place and make this component stateless instead. 
I'll change it.

File: src/components/elements/table/header.js (hunk 4 above)
Comment (robcar) [reply src/components/elements/table/header.js]: Yes, like @nranas said too right? Changed.

File: src/components/elements/table/header.js (hunk 3 above)
Comment (robcar) [reply src/components/elements/table/header.js]: Thanks, changed. I haven't quite got the hang of when you need them and when you don't.

File: src/components/elements/table/index.js (hunk 5 above)
Comment (robcar) [reply src/components/elements/table/index.js]: You're absolutely right. 
I tested with many rows and the performance gain with regex was too big. 
Changed

File: src/components/elements/table/header.js [hunk 6]
```diff
@@ -1,53 +1,35 @@
...
-    })
-    this.props.sortBy(columnIndex)
+export default ({ columns, sortedColumn, descending, sortBy }) => {
+  if (!columns) {
+    return <thead />
   }
+  return (
+    <thead>
+      <tr className='ct-table-header-row'>
+        {columns.map((column, columnIndex) =>
```
Comment (hfjallemark) [inline src/components/elements/table/header.js:11]: Shouldn't there be a `(` on the end of this line (and `)` after the `</th>` closing tag)?

Comment (robcar) [reply src/components/elements/table/header.js:11]: I can sure add a set of extra parenthesis that renders fine. 

```
        {columns.map((column, columnIndex) => (
          <th
            onClick={e => {
              e.preventDefault()
              sortBy(columnIndex)
            }}
            key={column.id}
            className='ct-table-header'>
            <a href='#'
              className={'ct-table-header-sort-link'}>
              {column.title}</a>
            <span className={cx('ct-table-header-sort-direction', {
              'ct-table-header-sort-ascending': sortedColumn == columnIndex &&
              !descending,
              'ct-table-header-sort-descending': sortedColumn == columnIndex &&
              descending,
            })}>
            </span>
          </th>)
          )
        }
```
But this works equally well (at least in Chrome and Edge):
```
        {columns.map((column, columnIndex) => 
          <th
            onClick={e => {
              e.preventDefault()
              sortBy(columnIndex)
            }}
            key={column.id}
            className='ct-table-header'>
            <a href='#'
              className={'ct-table-header-sort-link'}>
              {column.title}</a>
            <span className={cx('ct-table-header-sort-direction', {
              'ct-table-header-sort-ascending': sortedColumn == columnIndex &&
              !descending,
              'ct-table-header-sort-descending': sortedColumn == columnIndex &&
              descending,
            })}>
            </span>
          </th>
          )
        }
```
Or am I missing something in the comment again?

Comment (hfjallemark) [reply src/components/elements/table/header.js:11]: Ah, yes I think you are right -- that works with arrow functions but not with a return statement. So you can't do this:

```javascript
function render() {
  return
    <th>
      ...
    </th>
}
```

But you can do this (as proven in this PR):

```javascript
const render = () =>
  <th>
    ...
  </th>
```

My personal taste would be to be consistent, but either way works.

Comment (robcar) [reply src/components/elements/table/header.js:11]: Cool, thanks for the lesson :)

**Review by nranas — APPROVED** [review]

File: src/components/elements/inputs/masked-input.js [hunk 7]
```diff
@@ -28,8 +28,7 @@ export default class MaskedInputInput extends Component {
   }
 
   adjustAndUpdate(value) {
-    this.state.displayValue = value
-    this.setState(state)
+    this.setState({ displayValue: value })
```
Comment (hfjallemark) [inline src/components/elements/inputs/masked-input.js:31]: Isn't this already done in master?

File: src/components/elements/inputs/masked-input.js [hunk 8]
```diff
@@ -4,7 +4,7 @@ import MaskedInput from 'react-text-mask'
 import createNumberMask from 'text-mask-addons/dist/createNumberMask.js'
 import emailMask from 'text-mask-addons/dist/emailMask.js'
 
-export default class MaskedInputInput extends Component {
+export default class MaskedInputField extends Component {
```
Comment (hfjallemark) [inline src/components/elements/inputs/masked-input.js:7]: This is unrelated to this PR, right? Do we still want to rename it? Just checking.

File: src/components/elements/table/index.js [hunk 9]
```diff
@@ -1,14 +1,73 @@
...
+    }
+  }
+
+  render() {
+    return (
+      <div className='ct-table-container'>
+        <label className='ct-element-label ct-table-label'>{this.props.property.title}</label>
+        <table className='ct-table'>
+          <Header
+            api={this.props.api}
```
Comment (hfjallemark) [inline src/components/elements/table/index.js]: I believe this prop is not being used anymore.

File: src/components/elements/inputs/masked-input.js (hunk 8 above)
Comment (robcar) [reply src/components/elements/inputs/masked-input.js:7]: Yes, @nranas correctly pointed out that it wasn't the best of namings.

**Review by hfjallemark — APPROVED** [review]

**Review by hfjallemark — APPROVED** [review]
